spec.html

```diff
@@ -6250,13 +6250,14 @@ <h1>CreateIntrinsics ( _realmRec_ )</h1>
         1. Set _realmRec_.[[Intrinsics]] to _intrinsics_.
         1. Let _objProto_ be ObjectCreate(*null*).
         1. Set _intrinsics_.[[%ObjectPrototype%]] to _objProto_.
-        1. Let _throwerSteps_ be the algorithm steps specified in <emu-xref href="#sec-%throwtypeerror%"></emu-xref> for the %ThrowTypeError% function.
-        1. Let _thrower_ be ! CreateBuiltinFunction(_throwerSteps_, &laquo; &raquo;, _realmRec_, *null*).
-        1. Set _intrinsics_.[[%ThrowTypeError%]] to _thrower_.
         1. Let _noSteps_ be an empty sequence of algorithm steps.
         1. Let _funcProto_ be ! CreateBuiltinFunction(_noSteps_, &laquo; &raquo;, _realmRec_, _objProto_).
         1. Set _intrinsics_.[[%FunctionPrototype%]] to _funcProto_.
-        1. Call _thrower_.[[SetPrototypeOf]](_funcProto_).
+        1. Let _throwerSteps_ be the algorithm steps specified in <emu-xref href="#sec-%throwtypeerror%"></emu-xref> for the %ThrowTypeError% function.
+        1. Let _thrower_ be ! CreateBuiltinFunction(_throwerSteps_, &laquo; &raquo;, _realmRec_, _funcProto_).
+        1. Perform ! DefinePropertyOrThrow(_thrower_, *"length"*, PropertyDescriptor { [[Value]]: *0*, [[Writable]]: *false*, [[Enumerable]]: *false*, [[Configurable]]: *false* }).
```

Comment:
This call to `DefinePropertyOrThrow` is odd: no-where in the current spec do we have an algorithm step that sets the `"length"` property of a built-in function. (And similarly for the `"name"` property.) As far as the algorithms are concerned, these properties are created by magic.

So while this step is (I believe) correct, a better approach might be to leave it magic for now, and have a separate PR that adds `_name_` and `_length_` parameters to `CreateBuiltinFunction`.

Reply:
We have SetFunctionLength, which I would use except for it defining the property as configurable.

Reply:
The algorithm steps (throwerSteps) usually define the length of the function; it shouldn't need to be explicitly specified.

Reply:
@ljharb: You're saying that a built-in function's algorithm steps usually define the function's length property? I'm pretty sure that's not the case. Do you have an example?

Reply:
@jmdyck when something is like

```
3.4.5 MyFunction ( a, b )
1. do x
2. do y
```

the length of the function is derived from the arguments in that declaration `a, b`. The only difference is that these functions are generally not explicitly created, they just exist magically by step 13 of CreateIntrinsics. There are a few cases like ThrowTypeError, promise resolve functions, etc, where they are explicitly created, and technically they never have a length set. I guess i can remove it from here and follow up in a separate pr.

Reply:
> @jmdyck when something is like
> 
> ```
> 3.4.5 MyFunction ( a, b )
> 1. do x
> 2. do y
> ```
> 
> the length of the function is derived from the arguments in that declaration `a, b`.

Yup, but:
- That value is derived from the clause heading, not the actual algorithm steps. If you pass "a set of algorithm steps" to `CreateBuiltinFunction`, it's not clear that that includes the clause heading. (It probably has to, in some sense, but it's not clear.)

- Even if we grant/decree that `CreateBuiltinFunction` can extract heading info from the `_steps_` parameter, the clause heading still only gives the *default* length. The actual value for the `"length"` property occasionally differs.

- And even if we grant (though this is a stretch) that `CreateBuiltinFunction` can extract the actual length value from the `_steps_` parameter, *it doesn't actually have a step that defines the `"length"` property*.

(Ditto that, more or less, for the `"name"` property.)

Reply:
I'm talking about functions that are created from "algorithm steps". See https://tc39.es/ecma262/#sec-promise-resolve-functions for example.

Reply:
I'm talking about those functions too (among others). Clause 17 tells us that promise resolve functions have a `"length"` property, and the clause you linked to tells us its value. But there is no algorithm step that actually creates/defines the function's `"length"` property when it's created in [`CreateResolvingFunctions`](https://tc39.es/ecma262/#sec-createresolvingfunctions).

Reply:
Fair, but making that implicitness more explicit is out of scope of this PR - this PR should specify the length along with the algorithm steps for the thrower function.

Reply:
> Fair, but making that implicitness more explicit is out of scope of this PR

Yes, exactly. That's what I suggested in the first place: "to leave it magic for now, and have a separate PR that adds `_name_` and `_length_` parameters to `CreateBuiltinFunction`."

> this PR should specify the length along with the algorithm steps for the thrower function.

You mean [9.2.9.1 %ThrowTypeError%](https://tc39.es/ecma262/#sec-%throwtypeerror%) should have a sentence like this?:
```
The "length" property of the %ThrowTypeError% function is 0.
```
Yeah, it could, but 0 is the default length that you'd infer from the clause heading anyway, and we don't usually have such a sentence in such cases.

Reply:
Yes, because these algorithm steps usually don’t/shouldn’t have a typical function signature, thus no length should be inferable.

Reply:
However, i do see how this one is a bit of a special case; what I’d expect instead of the current setup is something identical to https://tc39.es/ecma262/#sec-promise-resolve-functions

Reply:
A realm can have arbitrarily many promise resolve functions, but any clause heading that involved a parameter list would probably mislead the reader into thinking that there was only one such function per realm. Thus, the rule for inferring the value of the `"length"` properly doesn't apply, so the value of the property must be stated explicitly in the clause.

In contrast, there's exactly one `%ThrowTypeError%` function per realm (the same as all other intrinsics), so it's fine to put an (empty) parameter list in the clause heading, so the default length can be inferred. As long as the actual length value is the same as the inferred, no explicit statement is needed.

Reply:
I’m not sure why a clause header would imply one per realm at all; a function signature in code certainly doesn’t.

If there’s one per realm then i don’t know why we need the CreateBuiltinFunction stuff; other functions (like Array.isArray, say) don’t seem to need it.

Reply:
> I’m not sure why a clause header would imply one per realm at all

Convention, I guess. Every one-per-realm built-in function is declared with a parameter-list in its clause-heading, and every arbitrary-per-realm built-in function isn't. (There's one exception, which I think I'll submit a PR to fix.)

> If there’s one per realm then i don’t know why we need the CreateBuiltinFunction stuff; other functions (like Array.isArray, say) don’t seem to need it.

If, by "the CreateBuiltinFunction stuff", you mean steps 5-7 and 11 in the current `CreateIntrinsics`, then yes, I don't think we need it. I think it's just a leftover from when we actually did need it, or thought we did.